### PR TITLE
Check pending txs before submitting cancel

### PR DESCRIPTION
**Pull request by OGPoyraz**
## Explanation

This PR updates `speedUp` and `cancel` transactions to force checking pending transaction statuses before creating each transaction. 

## References

* Fixes https://github.com/MetaMask/metamask-extension/issues/22314

### `@metamask/transaction-controller`

- **Added**: Added a call of `PendingTransactionTracker.checkTransactions` before creating speed-up and cancel transactions

## Checklist

- [x] I've updated the test suite for new or updated code as appropriate
- [x] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [ ] I've highlighted breaking changes using the "BREAKING" category above as appropriate

### CODE REVIEW

File: packages/transaction-controller/src/TransactionController.test.ts
Comment (OGPoyraz) [inline packages/transaction-controller/src/TransactionController.test.ts:1875]: We are not simulating `update` (thought could be overkill to implement it just for this test) but rather checking if `checkTransactions` is called.

File: packages/transaction-controller/src/TransactionController.ts
Comment (matthewwalsh0) [inline packages/transaction-controller/src/TransactionController.ts:770]: The other downside of this is we're checking all pending transactions, so in the unlikely event you have 10 pending transactions for example, are we meaningfully delaying the submit and making this scenario more likely?

File: packages/transaction-controller/src/TransactionController.ts
Comment (OGPoyraz) [reply packages/transaction-controller/src/TransactionController.ts]: > From the controller perspective, we asked it to do something and it can't do it so it feels conceptually like an error.

I agree, but I have concerns if we only pass the error to the client and leave it to handle. (Assuming we don't want back and forth and prevent this happening `stopTransaction` => `error passed to client` => `client asks if tx is confirmed` => ... )

Also in case of an error, client could show some error stating that "this transaction is confirmed, you cannot cancel it". But if we don't force update of this transaction status, client will keep showing it as "submitted". 

In my opinion the best here is that we can send some error to the client(so it can show "too late" message) but also update, does it sound good on your end?

File: packages/transaction-controller/src/helpers/PendingTransactionTracker.ts
Comment (OGPoyraz) [reply packages/transaction-controller/src/helpers/PendingTransactionTracker.ts]: Thanks for underlining the second point, I totally missed that.

File: packages/transaction-controller/src/helpers/PendingTransactionTracker.ts
Comment (OGPoyraz) [inline packages/transaction-controller/src/helpers/PendingTransactionTracker.ts:152]: We could rename this if you find a better name

File: packages/transaction-controller/src/TransactionController.ts
Comment (matthewwalsh0) [inline packages/transaction-controller/src/TransactionController.ts:2686]: We've a pending ticket to remove the duplication between cancel and speed up, so I appreciate this method!

Perhaps a more explicit name would be `publishTransactionForRetry` so we know it's specific to speed up and cancel?

File: packages/transaction-controller/src/TransactionController.ts
Comment (matthewwalsh0) [inline packages/transaction-controller/src/TransactionController.ts:2696]: Is it worth us checking both the code itself but also the message content to cast a wider net?

File: packages/transaction-controller/src/TransactionController.ts
Comment (matthewwalsh0) [inline packages/transaction-controller/src/TransactionController.ts:2700]: Minor, and pending review from Barbara but no periods in single line messages.

File: packages/transaction-controller/src/TransactionController.ts
Comment (matthewwalsh0) [inline packages/transaction-controller/src/TransactionController.ts:2702]: For the sake of clarity, if it wasn't this already confirmed error, shall we throw the original error again so we don't impact more flows than intended?

File: packages/transaction-controller/src/helpers/PendingTransactionTracker.ts
Comment (OGPoyraz) [reply packages/transaction-controller/src/helpers/PendingTransactionTracker.ts]: You mentioned this before but I missed the point of globalLock, thanks for pointing out again. 
This is fixed in https://github.com/MetaMask/core/pull/3800/commits/2119ad1362caf48bd9d04c0f4877e6a4bdb76f15

File: packages/transaction-controller/src/TransactionController.ts
Comment (OGPoyraz) [reply packages/transaction-controller/src/TransactionController.ts]: Done in https://github.com/MetaMask/core/pull/3800/commits/27847911055229ab5ae0e3f748a2b0c741ed46e4

File: packages/transaction-controller/src/TransactionController.ts
Comment (OGPoyraz) [reply packages/transaction-controller/src/TransactionController.ts:2696]: Done in https://github.com/MetaMask/core/pull/3800/commits/c68f7fb7014444efe0ea97fd3471f937489e6871

Comment (OGPoyraz) [reply packages/transaction-controller/src/TransactionController.ts]: Done in https://github.com/MetaMask/core/pull/3800/commits/dc0a169ff0dc7574f1537ff6e67e57eb0bf0e08d

File: packages/transaction-controller/src/TransactionController.ts
Comment (OGPoyraz) [reply packages/transaction-controller/src/TransactionController.ts]: Good point, fixed in https://github.com/MetaMask/core/pull/3800/commits/84cf788287b53772c57b770eee99bc5818da3598

File: packages/transaction-controller/src/TransactionController.ts
Comment (OGPoyraz) [reply packages/transaction-controller/src/TransactionController.ts]: Fixed in https://github.com/MetaMask/core/pull/3800/commits/e0980caae73e868e350f724def200f65f83811d8

File: packages/transaction-controller/src/TransactionController.ts
Comment (matthewwalsh0) [inline packages/transaction-controller/src/TransactionController.ts:2685]: Rather than using a `let` here, could we instead just use a `const` and `return` it inside the `try` block?

File: packages/transaction-controller/src/TransactionController.ts
Comment (matthewwalsh0) [inline packages/transaction-controller/src/TransactionController.ts:2688]: Is it best practice if we avoid adding new usages of `any`?

Could this just be `unknown` so we can cast to the type we need inside `isNonceIssue`?

File: packages/transaction-controller/src/TransactionController.ts
Comment (matthewwalsh0) [inline packages/transaction-controller/src/TransactionController.ts:2695]: Minor, but do we need to check this text with Barbara?

File: packages/transaction-controller/src/utils/validation.ts
Comment (matthewwalsh0) [inline packages/transaction-controller/src/utils/validation.ts:75]: Is this technically part of validation as we're not checking any user inputs, rather just testing the type of error.

Would it be clearer inside `TransactionController` where it is used for now?

Comment (matthewwalsh0) [inline packages/transaction-controller/src/utils/validation.ts:75]: Perhaps `isTransactionAlreadyConfirmedError` to be explicit?

File: packages/transaction-controller/src/utils/validation.ts
Comment (matthewwalsh0) [inline packages/transaction-controller/src/utils/validation.ts:77]: Could `invalidInput` apply to other errors on submission unrelated to the confirmation already being confirmed?

Would we be safer just checking the error messages for now?

File: packages/transaction-controller/src/TransactionController.ts
Comment (OGPoyraz) [reply packages/transaction-controller/src/TransactionController.ts]: Done in https://github.com/MetaMask/core/pull/3800/commits/4006d719f526ca6ca796ebb7a50b859794fa9e6c

File: packages/transaction-controller/src/TransactionController.ts
Comment (OGPoyraz) [reply packages/transaction-controller/src/TransactionController.ts]: Done in https://github.com/MetaMask/core/pull/3800/commits/09b9e75b3d97c0f1fe14d88296f5ecfacc136f36

File: packages/transaction-controller/src/helpers/PendingTransactionTracker.ts
Comment (OGPoyraz) [reply packages/transaction-controller/src/helpers/PendingTransactionTracker.ts]: Done in https://github.com/MetaMask/core/pull/3800/commits/f55f9da2c54c5319ad99425660da1063aebb66ff

File: packages/transaction-controller/src/utils/validation.ts
Comment (OGPoyraz) [reply packages/transaction-controller/src/utils/validation.ts]: Done in https://github.com/MetaMask/core/pull/3800/commits/5a708d963a81bf9bf4f65d59dc742a56c0a45df8

Comment (OGPoyraz) [reply packages/transaction-controller/src/utils/validation.ts]: Done in https://github.com/MetaMask/core/pull/3800/commits/5a708d963a81bf9bf4f65d59dc742a56c0a45df8

File: packages/transaction-controller/src/utils/validation.ts
Comment (OGPoyraz) [reply packages/transaction-controller/src/utils/validation.ts]: Done in https://github.com/MetaMask/core/pull/3800/commits/5a708d963a81bf9bf4f65d59dc742a56c0a45df8

File: packages/transaction-controller/src/TransactionController.ts
Comment (OGPoyraz) [reply packages/transaction-controller/src/TransactionController.ts:2692]: https://consensys.slack.com/archives/C03RV0Q4TGB/p1707228944091439?thread_ts=1705657150.824489&cid=C03RV0Q4TGB

Comment (OGPoyraz) [reply packages/transaction-controller/src/TransactionController.ts:2692]: I am also considering to localise the copy here, but will handle in the extension I guess

**Comment by OGPoyraz** [comment]
@metamaskbot publish-preview

**Comment by OGPoyraz** [comment]
@metamaskbot publish-preview

**Review by matthewwalsh0 — APPROVED** [review]

**Review by vinistevam — APPROVED** [review]

**Pull request merged**
